### ISSUE TITLE
Visibility of running tasks

### ISSUE THREAD

**Issue by deboer-tim** [containers/podman-desktop#7400]
### Is your enhancement related to a problem? Please describe

#7399 will add a toast notification when a task is created, but we expect that either the user can close the toast or it will close automatically after a few seconds. This means that the user can't easily tell when it completes or if there is still a task running without opening the task manager.

### Describe the solution you'd like

We need some visible indication when there is a task running. Some options that have been raised:
- animation of the current bell
- adding the number of running tasks to the bell
- status bar entry with the name of the currently running task (or some flavour of '2 tasks running')
- progress bar or spinner directly on the status bar

**Comment by benoitf** [comment]
it may overlap with https://github.com/containers/podman-desktop/issues/8546

**Comment by axel7083** [comment]
> animation of the current bell

This does not seems to be a common pattern ? I cannot really find much examples of how it could be animated.

We have the the dot to highlight, I have some understanding that it may not be visible enough but we are limited by the size of status bar.

![image](https://github.com/user-attachments/assets/3132bf0b-3cdb-4e01-ae94-0dff9bf80973)

> adding the number of running tasks to the bell

As per the previous comment, with the current status bar size, I feel like it would be noisy to add a number as it is very small.

> status bar entry with the name of the currently running task (or some flavour of '2 tasks running')
> progress bar or spinner directly on the status bar

Those two ideas seems the same. And I like it as it is the way my Jetbrain IDE is doing

https://github.com/user-attachments/assets/e635871b-3021-40ff-9105-dfdec1a71caf

![image](https://github.com/user-attachments/assets/467bede7-7bc3-4167-be99-43c84acc6274)

We can split it in three, the name, the loading, and the action ? Maybe we should limit to cancellation action. 

## Partial Solution proposal

For a first we could have a basic progress bar at the bottom as suggested

https://github.com/user-attachments/assets/ac5c921a-a6bc-4372-b7d0-e3b16a9b6ebb

We could make it clickable and open the task manager

**Comment by slemeur** [comment]
> As per the previous comment, with the current status bar size, I feel like it would be noisy to add a number as it is very small.

It really depends, if the tool is firing dozen of notifications I agree. But as a user, if a task is failing, or if I have an error reported, it should caught my attention.
I think we should think about a model when the status bar is highlighting the case when there is an error with one of the tasks/notification.

The proposed solution is interesting, a couple of feedback:
- the "in progress operation" in the status bar should be closer to the notification icon, otherwise it feels disconnected. 
- I think it should be displayed in the status bar, only when the task manager is folded. When unfolded, that feels overloading the UI with a lot of animated loaders.

**Comment by deboer-tim** [comment]
I really like the proposal (especially since it matches a hack I tried months ago 😉). It's obvious when something is running in the background and handles multiple running tasks fine. IMHO the status progress bar could be smaller, and it's ok at least for now to not add any actions (i.e. you have to open the task manager to do anything).

> the "in progress operation" in the status bar should be closer to the notification icon, otherwise it feels disconnected.

Yes, this is something we need to resolve. We could move the task manager icon to the left of the version, but that looks a little out of place since then you have icon-text-icon-icon the rest of the time. @ekidneyrh any suggestions?

> I think it should be displayed in the status bar, only when the task manager is folded. When unfolded, that feels overloading the UI with a lot of animated loaders.

I checked a couple similar tools and they don't remove it - but I think it works better in those cases because the status bar is a fraction of the size so the 'double progress' doesn't stand out so much. Maybe we can start by making it 50% smaller and see if that helps?

**Comment by axel7083** [comment]
> It really depends, if the tool is firing dozen of notifications I agree. But as a user, if a task is failing, or if I have an error reported, it should caught my attention.
I think we should think about a model when the status bar is highlighting the case when there is an error with one of the tasks/notification.

We are suppose to have the toast for that, here is an example in JetBrain products

![image](https://github.com/user-attachments/assets/2878cf00-ad01-4d69-88e8-a39004c0f6cb)

Another example in VSCode

![image](https://github.com/user-attachments/assets/90e950e9-d039-4eb4-8f07-ab866cb5d415)

I might be wrong, but that's not a pattern I see in any tools, if you have references where the statusbar show the error in some way it might help have a better understanding of how it can be shown.

IMO if a task fail, it should create a notification (as in vscode) with an error associated or something. But the status bar should not have the responsibility of showing the error. 

> the "in progress operation" in the status bar should be closer to the notification icon, otherwise it feels disconnected.

I also agree

> I think it should be displayed in the status bar, only when the task manager is folded. When unfolded, that feels overloading the UI with a lot of animated loaders.

vscode and jetbrain keep it, even tho it may feel a bit redundant. 

> Maybe we can start by making it 50% smaller

Do you mean in width ?

**Comment by deboer-tim** [comment]
> IMO if a task fail, it should create a notification (as in vscode) with an error associated or something. But the status bar should not have the responsibility of showing the error.

I agree - status bar should just be a minimal reminder that something is running in the background, failures should use toast notification.

> > Maybe we can start by making it 50% smaller
> 
> Do you mean in width ?

Yes. Wide enough that you can still see some progress/animation, but small enough that it's not visually distracting or similar to what's in the task manager.

**Comment by slemeur** [comment]
>>IMO if a task fail, it should create a notification (as in vscode) with an error associated or something. But the status bar should not have the responsibility of showing the error.

>I agree - status bar should just be a minimal reminder that something is running in the background, failures should use toast notification.

On this comment - do you envision the toast to disappear after some time?
If there is an error or a miss configuration which needs action from the user, it's highly possible that the toast will be missed and the user will never check the "task" manager if there is no long lasting attention drawn to the end-user. When you are building your project in an IDE, you can see the error in your status bar and you have a quick way to access it (even if you have the toast). Like here:
![3KpUL](https://github.com/user-attachments/assets/57adf6e7-cf38-4dc1-a818-661da05cec34)

>>Maybe we can start by making it 50% smaller
Do you mean in width ?
Yes. Wide enough that you can still see some progress/animation, but small enough that it's not visually distracting or similar to what's in the task manager.

Let's please use a different loader style with something more discrete. A spinner would be lighter and more elegant.
![images](https://github.com/user-attachments/assets/c3c3325c-c210-4eb9-a5a4-3baa3b23c6c6)

**Comment by axel7083** [comment]
> On this comment - do you envision the toast to disappear after some time?

I don't know I would say that's a possibility, but I am not a UX :)

> Let's please use a different loader style with something more discrete. A spinner would be lighter and more elegant.

What do you think @deboer-tim

**Comment by deboer-tim** [comment]
> > On this comment - do you envision the toast to disappear after some time?
> 
> I don't know I would say that's a possibility, but I am not a UX :)

If you do toasts (e.g. VS Code) then you see a one line description of the error and they generally stay until you click for more details or close them. I'm ok just doing a status bar error instead, and (as today) you open the task manager for details.

Either way we should track via a separate issue since it is not related to this one - e.g. status bar error is changing the existing task manager icon to an error marker and can be done independent of this issue.

> > Let's please use a different loader style with something more discrete. A spinner would be lighter and more elegant.
> 
> What do you think @deboer-tim

We have tasks with progress (5%, 27%, ...) and indefinite tasks (don't know when they will be done). @slemeur I'm fine with it being styled differently than the task manager, but I'd really like to keep something that shows progress for tasks that have it. We could try an animated spinner for indefinite and do a slow build of the spinner when there is progress, but  at the size I'm worried there's not much detail to build from.

**Comment by axel7083** [comment]
> I'm ok just doing a status bar error instead, and (as today) you open the task manager for details.

We just show in several tools this is not the how it is done, I am not sure to understand why we are going back ?

> We have tasks with progress (5%, 27%, ...)

I think, for most of tasks (download something) we have a loading bar with progress, and user probably want to be able to see it.

If multiples running tasks are loading in parallel I think an interdeterminate would be more appropriate as it keep some consistency, instead of going from loading progress bar to spinner ?

> but at the size I'm worried there's not much detail to build from.

For a progress bar, this example I show above would be clear enough, even a bit smaller, but a spinner, or circlar progress indicator, it would not

**Comment by ekidneyrh** [comment]
I agree with @deboer-tim  that the task failing should be a different issue. 

Re-organising the task bar should be easy enough visually. We can move the version number to the far right and then have the bell icon on the far left so it’s next to whatever progress element we put there.

![Screenshot from 2024-09-13 12-22-00](https://github.com/user-attachments/assets/d691702b-a352-4c2c-9657-b4085010e0dc)


In AI Lab we use the progress spinner:

![Screenshot from 2024-09-13 10-54-35](https://github.com/user-attachments/assets/eb0b3280-d497-4fab-9c1e-38b30fdfc86b)



It’s indeterminate but for the task bar we can have it be determinate if the tasks are longer loading to show the user the progress being made. 

Using a bar could look like this:

![Screenshot from 2024-09-13 12-31-57](https://github.com/user-attachments/assets/af674027-938b-4880-8644-a9ed76b306d9)

**Comment by Firewall** [comment]
That looks great! Could we re-use the same pattern and area for failed tasks. So that would be an `Task failed ❌` and text in red.

**Comment by benoitf** [comment]
what would be the value of the progress bar when we have 2 tasks running ?
![image](https://github.com/user-attachments/assets/5e9481e3-0696-438b-aea9-394bb5fede97)

is it the average percent of the two tasks ?

**Comment by deboer-tim** [comment]
I agree any progress _value_ wouldn't be obvious, but IMHO using the indeterminate progress bar would keep the UI more consistent (wouldn't change as much when a second task is started or ends) and keeps a little visual reminder that there is something running in the background.

**Comment by axel7083** [comment]
> what would be the value of the progress bar when we have 2 tasks running ?

indeterminate

**Comment by axel7083** [comment]
> Could we re-use the same pattern and area for failed tasks. So that would be an `Task failed ❌` and text in red.

I would take example on VSCode and Jetbrain on this one, and avoid adding completed (either failed or successful) task in the status bar, there is the TaskManager for it.

Failed tasks would be displayed and visible through the toast (see  https://github.com/containers/podman-desktop/issues/7400#issuecomment-2342994109)

**Comment by slemeur** [comment]
The goal of this issue is to raise awareness to the end-user about potential long lasting tasks (downloads of a model for example). 
Creating a separate issue for the failed tasks: #9059

To summarize the different ideas:

### Showing visual progress indicator
This is something that everybody is aligned with.
We'll need to establish the rules for the following cases:
- an indeterminate task
- an additional tasks added while a determinate task is already running
- an additional tasks added while a indeterminate task is already running

### Style for visual progress indicator
I think the horizontal progress is taking quite some space. Could it be a circle that is showing progress in the status bar. Like here, but still showing progress:
![367324336-5e9481e3-0696-438b-aea9-394bb5fede97](https://github.com/user-attachments/assets/bd6ebcfd-4875-4aaa-87fc-5b73ed18cf37)

### Showing a label in the status bar next
The example from IntelliJ is pretty neat, but the label is actually very descriptive about the tasks being executed. 
- Would we be able to provide that level of details for the label?

### Showing label for number of tasks running
We all agreed that we need to show how many tasks are currently running. 
I'm wondering if that is really a necessity to show a label like "2 tasks are running". It feels not giving a lot of value. 
If we are moving to a circle progress indicator, maybe we could display the number of tasks in progress into the circle?

### Status bar change for the tasks
I believe, we need to adjust the status bar, to be able to display the "running tasks" properly. Should we think about changing the "bell" icon to something more representative?
<img width="172" alt="Screenshot 2024-08-22 at 11 47 24" src="https://github.com/user-attachments/assets/d31e28b5-09ea-4dda-af5a-2ea2cec17022">

@ekidneyrh I think we need to detail a couple of cases:
1. What do we display when nothing is running?
2. What do we display when there is one tasks in progress
3. What do we display when there is multiple tasks in progress
Could you please explore this?

**Comment by ekidneyrh** [comment]
> What do we display when nothing is running?

I think if nothing is running, just have nothing be there? I am presuming that the bell icon is still there if the user wants to access the task manager still. 

> What do we display when there is one tasks in progress

It depends on what the task is. I think displaying what the task it would be beneficial but we can limit the characters that are used so it’s not a super long message. To do this we can see what the current tasks are and shorten the output - this will probably be an iterative process as we will probably miss some. 

> What do we display when there is multiple tasks in progress
> Could you please explore this?

Having the ‘x tasks running’ is what was decided upon in yesterday's meeting. 

I can have a look at what icons would be an alternative to the bell if needed- maybe something like GitLab has for their assigned issues or Githubs notifications?


![Screenshot from 2024-09-26 10-05-11](https://github.com/user-attachments/assets/a0c82c2b-9b81-45cd-9882-79f3cc126a35)


![Screenshot from 2024-09-26 10-13-11](https://github.com/user-attachments/assets/cf91ae87-fadf-42a5-be8a-0c697d49a202)
